dvc/remote/__init__.py

```diff
@@ -90,3 +72,17 @@ def _resolve_remote_refs(config, remote_conf):
     base = config["remote"][parsed.netloc]
     url = posixpath.join(base["url"], parsed.path.lstrip("/"))
     return {**base, **remote_conf, "url": url}
+
+
+def get_cloud_tree(repo, **kwargs):
+    remote_conf = _get_conf(repo, **kwargs)
+    return _get_tree(remote_conf)(repo, remote_conf)
+
+
+def get_remote(repo, **kwargs):
+    tree = get_cloud_tree(repo, **kwargs)
+    if tree.scheme == "local":
+        return LocalRemote(tree)
+    if tree.scheme == "ssh":
+        return SSHRemote(tree)
+    return Remote(tree)
```

Comment:
For initializing caches, we now just do
```
tree = get_cloud_tree(<remote config>)
cache = CloudCache(tree)
```

But for remotes, we still have this helper method. It would be ideal if we could just always use `Remote(tree)` but `LocalRemote` and `SSHRemote` still have their own overridden batch exists methods. It may be worth investigating whether or not the custom ssh/local methods are still needed with the other performance improvements that we have now.

Reply:
Or maybe those custom methods could be moved to the trees. IIRC ssh uses some custom sftp channel handling, which might be abstracted away too. But surely could do that later. Maybe consider creating an issue for it.

Reply:
One other question I had was whether or not the indexing related code should also just go into the trees at some point instead of being DVC Remote specific

Reply:
@pmrowla Oh, that's a very good question! To me, it seems purely remote-related, as it solely relies on remote structure. But how do you see it?

Reply:
Since it's remote specific it makes sense to keep it in the Remote class for now, but I wasn't sure if indexing any kind of content (as opposed to just DVC Remotes) in a remote/cloud tree is something we might want to support in the future.

Reply:
I'm not sure there is a good way to index arbitrary data in that way, it is really tailored for our  cache structure, so I really doubt that full index logic will ever make its way into the Tree. But, I'm sure that parts of it will, as many functions are actually quite generic (like walk_files replacing _list_cache).